File: projects/project_69/src/analysis/analysis.py
#!/usr/bin/env python
# Title: Data Science DSC180A (Replication Project), DSC180B (miRNA Overlap Alzheimer's and Parkinson's)
# Section B04: Genetics
# Authors: Saroop Samra (180A/180B), Justin Kang (180A), Justin Lu (180B), Xuanyu Wu (180B)
# Date : 10/23/2020

# This step will take the preprocessed fastq data and analyze the RNA levels for the 4 groups and 
# determine correlations

import os
import re
import pandas as pd
import numpy as np
import logging
    


def process_biofluid_disorder_comparisons(deseq2, output_prefix, cleanup, verbose):
    '''
    Purpose : Analyze the disorder comparisons per brain region
    Arguments: 
    deseq2          : dictionary of normalize deseq2 config arguments
    output_prefix   : output prefix for all files (normalized matrix, LRT, etc)
    cleanup         : if set to 1 then remove merged bam input file
    verbose : set to 1 if you want verbose output
    '''
    tmp_files = []
    # Iterate for each brain region
    for biofluid_region in deseq2["biofluid_regions"]:
        # Create Directory
        biofluid_dir = output_prefix.replace("%biofluid_region%", biofluid_region) + "/"
        if not os.path.isdir(biofluid_dir):
            os.makedirs(biofluid_dir)
            if verbose:
                logging.info("mkdir " + biofluid_dir)
        # load full matrix and features
        biofluid_count_matrix_df = pd.read_csv(deseq2["input_counts"], sep="\t", index_col=0)
        biofluid_feature_df = pd.read_csv(deseq2["input_features"], sep="\t")
        # filter brain region
        biofluid_feature_df = biofluid_feature_df[biofluid_feature_df["Biofluid"] == biofluid_region]


        for disorder in deseq2["disorders"]:
            disorder_a =  disorder
            disorder_b =  deseq2["control"]
            disorder_dir = (biofluid_dir + disorder).replace(" ", "_") + "/"
            if not os.path.isdir(disorder_dir):
                os.makedirs(disorder_dir)
                if verbose:
                    logging.info("mkdir " + disorder_dir)


            feature_df = biofluid_feature_df.copy()
            count_matrix_df = biofluid_count_matrix_df.copy()

            logging.info(biofluid_region + " x " + disorder_a + " vs Control")
            # filter all a and b disorders
            feature_df = feature_df[ (feature_df["Disorder"] == disorder_a) | (feature_df["Disorder"] == disorder_b)] 
            feature_df["Disorder"] = feature_df["Disorder"].replace({disorder_a:0, disorder_b:1})

            # filter samples in count matrix
            count_matrix_df = count_matrix_df[ feature_df["Run"].tolist() ]
            # save the filtered count and feature tables
            input_count_filename = disorder_dir + "gene_matrix.tsv"
            tmp_files.append(input_count_filename)
            count_matrix_df.to_csv(input_count_filename, sep="\t")
            input_feature_filename = disorder_dir + "features.tsv"
            tmp_files.append(input_feature_filename)
            
            # remove biofluid_region from feature table
            feature_df = feature_df.drop(["Biofluid"], axis=1) # , "Run"
            feature_df.to_csv(input_feature_filename, sep="\t", index=False)

            # DEQSe2 Process
            if os.path.exists(deseq2["Rscript"]):
                command = deseq2["Rscript"] + " "
            else:
                command = "Rscript "
            command += deseq2["source"]
            command += " "
            command += input_count_filename
            command += " "
            command += input_feature_filename
            command += " "
            command += disorder_dir
            command += " full="
            command += str(deseq2["full"])
            command += " reduced="
            command += str(deseq2["reduced"])
            command += " charts=1"
            if deseq2["parallel"]==1:
                command += " parallel=1"
            else:
                command += " parallel=0"
            if verbose:
                logging.info(command)
            os.system(command)

            if cleanup == 1:
                for tmp_file in tmp_files:
                    if verbose:
                        logging.info("rm " + tmp_file)
                    os.remove(tmp_file)   
                tmp_files = []





def process_analysis(deseq2, output_prefix, cleanup, verbose):
    '''
    Purpose : normalize merged BAM script using R's DESeq2
    Arguments: 
    deseq2          : dictionary of normalize deseq2 config arguments
    output_prefix   : output prefix for all files (normalized matrix, LRT, etc)
    cleanup         : if set to 1 then remove merged bam input file
    verbose : set to 1 if you want verbose output
    '''
    
    if verbose:
        logging.info("# ---------------------------------------------------")
        logging.info("# Analysis")
    
    # process each brain region x disorder - used in pvalue histogram plot
    process_biofluid_disorder_comparisons(deseq2, output_prefix, cleanup, verbose)


    if verbose:
        logging.info("# Finished")
        logging.info("# ---------------------------------------------------")

File: projects/project_69/src/analysis/analysis.r
#!/usr/bin/env Rscript

# Title: Data Science DSC180A (Replication Project), DSC180B (miRNA Overlap Alzheimer's and Parkinson's)
# Section B04: Genetics
# Authors: Saroop Samra (180A/180B), Justin Kang (180A), Justin Lu (180B), Xuanyu Wu (180B)
# Date : 10/23/2020


# MAIN CITED PAGE:
# Cite: https://bioinformatics.uconn.edu/resources-and-events/tutorials-2/rna-seq-tutorial-with-reference-genome/#

# Add this to use the R from RStudio which has faster DESeq2
.libPaths( "/opt/conda/envs/r-bio/lib/R/library" )


# check if package not installed
#if (!requireNamespace("BiocManager", quietly = TRUE)) {
#    install.packages("BiocManager")
#	BiocManager::install("DESeq2")
#}

library("RColorBrewer")
#if (!requireNamespace("gplots", quietly = TRUE)) {
#    install.packages('gplots',repos = "http://cran.us.r-project.org")
#}
suppressPackageStartupMessages( library("gplots") )

# load library but don't show messages
suppressPackageStartupMessages( library( "DESeq2" ) )

# Parallel Processing to make it faster
library("BiocParallel")

# Print the DESeq version
print("DESeq Package version=")
packageVersion("DESeq2")

# get the command line arguments
args = commandArgs(trailingOnly=TRUE)

# verify args
if (length(args) < 7) {
	stop("Missing arguments : <gene matrix file> <experiment file> <output_prefix> <full> <reduced> charts=[0|1] parallel=[0|1]")
}

gene_matrix_file = args[1]
experiment_file = args[2]
output_prefix = args[3]
full = args[4]
reduced = args[5]
charts = args[6]
parallel = args[7]

# convert string to variables we can pass to R
# Cite: https://win-vector.com/2018/09/01/r-tip-how-to-pass-a-formula-to-lm/
# full <- as.forumla(paste(full, collapse = " + "))



# I used the following example to base my code on
# Cite: https://dputhier.github.io/ASG/practicals/rnaseq_diff_Snf2/rnaseq_diff_Snf2.html

# Load to tables
experiment_features <- read.table(file=experiment_file, sep="\t", header=TRUE)
gene_count.table <- read.table(file=gene_matrix_file, sep="\t", header=TRUE, row.names=1)
print("Loading Complete")


# Deal with NaN's: Now done in Pandas
#gene_count.table[is.na(gene_count.table)] <- 0

# TODO: PRUA, percentage of genes uniquely aligned
# TODO : Remove Non-convergent genes
# TODO: correct for PRUA by computing residuals to linear model regression PRUA on nornalized gene...

## Use the DESeqDataSetFromMatrix to create a DESeqDataSet object
#deseq_data <- DESeqDataSetFromMatrix(countData = gene_count.table, colData = experiment_features, design = ~ Age+PMI+pH+PRUA+Disorder)
# Add 1 to avoid error: normalize: every gene contains at least one zero, cannot compute log geometric 

print(paste("gene count dimensions", dim(gene_count.table)))
print(paste("experiment features dimensions", dim(experiment_features)))

deseq_data <- DESeqDataSetFromMatrix(countData = round(gene_count.table+1), colData = experiment_features, design = ~ expired_age+sex+PMI+sn_depigmentation+Braak_Score+TangleTotal+Plaque_density+PlaqueTotal+Disorder)
print("Matrix Set Complete")

dds_norm <-  estimateSizeFactors(deseq_data)
print("estimateSizeFactors complete")

# Likelihood Ratio Test (LRT)
if (parallel == "parallel=1") {
	# Parallel version
	dds_lrt <- DESeq(deseq_data, test="LRT", parallel=TRUE, BPPARAM=MulticoreParam(workers=4), reduced =~ expired_age+sex+PMI+sn_depigmentation+Braak_Score+TangleTotal+Plaque_density+PlaqueTotal)
} else {
	dds_lrt <- DESeq(deseq_data, test="LRT", reduced =~ expired_age+sex+PMI+sn_depigmentation+Braak_Score+TangleTotal+Plaque_density+PlaqueTotal) 
}

print("DESeq LRT setup complete")  


# Extract results
res_LRT <- results(dds_lrt)

print("DESeq LRT results complete")

# Save LRT results
write.table(res_LRT, file=paste(output_prefix, "lrt.tsv", sep = ""), sep="\t", quote=F, col.names=NA)

print("DESeq LRT saving complete")

if (charts == "charts=1") {
	# Output MAPlot
	png(paste(output_prefix, "MAplot.png", sep = ""))
	plotMA(res_LRT, ylim=c(-8,8),main = "")
	dev.off()

	print("DESeq MA Plot complete")

	# Variance Stabilizing Transformation (VST)
	vsd <- varianceStabilizingTransformation(deseq_data, blind=T)

	print("DESeq VST Setup complete")

	# Heatmap of data of VST with 1000 top expressed genes with heatmap.2
	select <- order(rowMeans(counts(dds_lrt,normalized=T)),decreasing=T)[1:20]
	my_palette <- colorRampPalette(c("blue",'white','red'))(n=20)
	png(paste(output_prefix, "heatmap.png", sep = ""))
	heatmap.2(assay(vsd)[select,], col=my_palette,
	          scale="row", key=T, keysize=1, symkey=T,
	          density.info="none", trace="none",
	          cexCol=0.6, labRow=F,
	          main="")
	dev.off()

	print("DESeq Heatmaps complete")

}


